libs/brig-types/src/Brig/Types/Common.hs

```diff
@@ -386,3 +386,44 @@ codeParser :: String -> (String -> Maybe a) -> Parser a
 codeParser err conv = do
     code <- count 2 anyChar
     maybe (fail err) return (conv code)
+
+-----------------------------------------------------------------------------
+-- ManagedBy
+
+-- | Who controls changes to the user profile (where the profile is defined as "all
+-- user-editable, user-visible attributes").
+data ManagedBy
+      -- | The profile can be changed in-app; user doesn't show up via SCIM at all.
+    = ManagedByWire
+      -- | The profile can only be changed via SCIM, with several exceptions:
+      --
+      --   1. User properties can still be set (because they are used internally by clients
+      --      and none of them can be modified via SCIM now or in the future).
+      --
+      --   2. Password can be changed by the user (SCIM doesn't support setting passwords yet,
+      --      but currently SCIM only works with SSO-users who don't even have passwords).
+      --
+      --   3. The user can still be deleted normally (SCIM doesn't support deleting users yet;
+      --      but it's questionable whether this should even count as a /change/ of a user
+      --      profile).
+      --
+      -- There are some other things that SCIM can't do yet, like setting accent IDs, but they
+      -- are not essential, unlike e.g. passwords.
+    | ManagedBySCIM
+    deriving (Eq, Show, Bounded, Enum)
+
+instance FromJSON ManagedBy where
+    parseJSON = withText "ManagedBy" $ \case
```

Comment:
Is it reasonable to pull out a helper:

```haskell
typeName :: forall a. Typeable a => String
typeName = show $ typeRep @a
```

And use something like 

```haskell
withText (typeName @ManagedBy)
```
So we get a compile-time error here if the type is renamed?

I suppose this should all be covered by tests anyways; probably not worth it 🤷‍♂️

Reply:
I don't think `typeName` will compile without a `Proxy` argument, but then it would work, and I am in favor of the idea.

Reply:
@fisx We would need `AllowAmbiguousTypes`; not sure if the team has opinions on that particular extension or not 🤷‍♂️

Reply:
@neongreen Or am I misunderstanding the purpose of this string; does it need to match the type name exactly or is it actually just for error messages?

Reply:
> I don't think `typeName` will compile without a `Proxy` argument

It will if you enable AllowAmbiguousTypes (which you should)

Reply:
This said, the type name doesn't really matter. It only shows up in Aeson error messages and I'm not sure that deviating from this "standard" way of writing JSON instances is worth it.